### PR TITLE
Add PyTorch and CUDA memory profiling systems

### PR DESCRIPTION
**Pull request by aospan**
#  Why?
To gain deep visibility into PyTorch execution and GPU memory behavior, enabling faster debugging, performance tuning, and better understanding of CUDA memory usage - ideal for both practical troubleshooting and educational exploration.

# Profiling Guide for Nanochat Training

## Overview

The profiling system supports two types of profiling:
1. **PyTorch Profiler** - Captures detailed CPU/CUDA activity traces and memory timelines
2. **CUDA Memory Snapshot** - Captures detailed memory allocation/deallocation events

## Quick Start

To enable profiling, add `--enable_profiling=True` to the `scripts.base_train` command in `speedrun.sh` e.g.:

```bash
torchrun --standalone --nproc_per_node=$NUM_GPU scripts.base_train --enable_profiling=True
```

This enables both PyTorch profiler and CUDA memory profiling for all phases. Look for cyan-colored `[PROFILER:phase_name]` messages in the output to track profiling progress.

## Output Files

Profiling outputs are saved to `<base_dir>/profile_traces/<timestamp>/` where each run gets its own timestamped subdirectory to avoid overwriting previous profiles (e.g. `~/.cache/nanochat/profile_traces/20250117_143022/`).

### PyTorch Profiler Outputs
- `{stage}-{phase_name}_trace.json` - Chrome trace file (viewable in chrome://tracing or https://ui.perfetto.dev/)
- `{stage}-{phase_name}_memory_timeline.html` - Memory timeline visualization

### CUDA Memory Snapshot Outputs
- `{stage}-{phase_name}_mem.pickle` - Memory snapshot (can be analyzed by visiting https://docs.pytorch.org/memory_viz)

**Stage prefixes**: Files are prefixed with `stage1-`, `stage2-`, `stage3-`, `stage4-` to indicate the order of profiling phases:
- `stage1-` = model_init (model creation, weight init, compilation)
- `stage2-` = eval_bpb (first validation evaluation)
- `stage3-` = training_microsteps (training micro-steps)
- `stage4-` = optimizer_step (gradient clipping, learning rate scheduling, optimizer step, zero_grad)

### Example Output Files

Typical profiling run generates the following files (sizes may vary based on model size and profiling duration):

```bash
$ cd ~/.cache/nanochat/profile_traces/20251018_111748
$ du -hs *
100K    stage1-model_init_mem.pickle
56K     stage1-model_init_memory_timeline.html
64M     stage1-model_init_trace.json
53M     stage2-eval_bpb_mem.pickle
52K     stage2-eval_bpb_memory_timeline.html
73M     stage2-eval_bpb_trace.json
43M     stage3-training_microsteps_mem.pickle
236K    stage3-training_microsteps_memory_timeline.html
126M    stage3-training_microsteps_trace.json
1.2M    stage4-optimizer_step_mem.pickle
76K     stage4-optimizer_step_memory_timeline.html
369M    stage4-optimizer_step_trace.json
```

**Total**: ~728 MB for a complete profiling run with all phases enabled.

### Chrome Traces
1. Open a web browser and visit https://ui.perfetto.dev/
2. Click "Open trace file" and select a `*_trace.json` file
3. Explore the timeline view with zoom and pan

profiling-micro-steps-traces

*Nanochat profiling results: Training microsteps trace showing CPU/CUDA activity timeline down to individual CUDA kernel calls*

### Memory Timelines
1. Open the `*_memory_timeline.html` file in a web browser
2. Explore memory allocation patterns over time

profiling-micro-steps-memory-timeline

*Nanochat profiling results: Memory timeline visualization showing allocation patterns across training micro-steps*

### CUDA Memory Snapshots
1. Visit https://pytorch.org/memory_viz in a web browser
2. Upload the `*_mem.pickle` file to analyze memory patterns

profiling-micro-steps-cuda-memory-allocations

*Nanochat profiling results: CUDA memory snapshot showing detailed memory allocations by category*

### CODE REVIEW

**Comment by aospan** [comment]
Here’s one of the traces captured during nanochat training on my GPU. There are no gaps between CUDA kernel executions - meaning the GPU isn’t idling. The green “Command Buffer Full” marker also shows that the CPU is issuing CUDA kernels and API calls faster than the GPU can process them, which further confirms the GPU is fully utilized.

cuda_full

**Comment by aospan** [comment]
Perfetto allows executing SQL queries to get more insights from profiling data, so I ran this query to get the top 10 CUDA kernels executed during nanochat training steps. These kernels together account for ~45% of total GPU time.

Below is the SQL query and a screenshot of the results:
```
SELECT
  COUNT(dur) AS call_count,
  SUM(dur) / 1000000 AS total_duration_ms,
  AVG(dur) / 1000000 AS avg_duration_ms,
  (SUM(dur) * 100.0 / (
    SELECT SUM(dur)
    FROM slice
    WHERE track_id IN (3)
      AND name IS NOT NULL
      AND name NOT LIKE 'ProfilerStep#%'
  )) AS duration_percent,
  name
FROM slice
WHERE track_id IN (3)
  AND name IS NOT NULL
  AND name NOT LIKE 'ProfilerStep#%'
GROUP BY name
ORDER BY duration_percent DESC
LIMIT 10;
```

Screenshot 2025-10-19 at 16 40 29


First two (`ampere_bf16_s1688gemm_bf16_128x128_ldg8_f2f_stages_32x1_nn` and `ampere_bf16_s16816gemm_bf16_256x128_ldg8_f2f_stages_32x3_nn`) are BF16 precision CUTLASS GEMM (general matrix multiply) ops running on Ampere tensor cores (yep, experimenting on a humble RTX 3060 12GB). The `128x128` tile means each thread block computes a 128×128 output tile. `s1688gemm` dictates how the input matrices are partitioned for computation. `ldg8` Indicates the use of 8-wide vectorized loads, optimizing memory access patterns. `f2f` - fused type-casting between different floating-point data types. `stages_32x1` - 32 stages and a 1x1 threadblock cluster. `nn` - matrix multiplication layout (both matrices are in "normal" (non-transposed)).

For further reading, please dig into the CUTLASS repo here: https://github.com/NVIDIA/cutlass - and may the force be with you :)

**Review by svlandeg — COMMENTED** [review]
Hi @aospan! 

These kind of profiling systems can definitely be useful, but the diff shows hows this functionality interferes with the main code in `base_train.py`. I think it obfuscates the core training loop, which is supposed to be minimalistic for educational purposes. As such, I think we can close this one.

**Comment by aospan** [comment]
@svlandeg  Thanks for the review and feedback! I agree - this adds extra steps to the training loop and might be a burden for those who learning.

Ideally, we’d start the profiler at the beginning and stop it at the end of the entire program, but that produces a huge amount of data and can cause the process to freeze. That’s why I limited profiling to the first ~10 training loops - after that, they mostly repeat and add little new value.

I’m fine with leaving this unmerged; anyone interested can still apply the patches locally. Could even be a neat addition to @karpathy 's courses - a module on performance tuning down to CUDA kernels and hardware.

**Comment by karpathy** [comment]
Definitely a good patch to have for the repo but yeah struggling to see how to neatly include it into master without bloating the code. Ty!
